Sources/SwiftLexicalLookup/SimpleLookupQueries.swift

```diff
@@ -0,0 +1,176 @@
+  private func lookupCatchNodeHelper(at syntax: Syntax?, traversedCatchClause: Bool) -> Syntax? {
+    guard let syntax else { return nil }
+
+    switch syntax.as(SyntaxEnum.self) {
+    case .doStmt:
+      if traversedCatchClause {
+        return lookupCatchNodeHelper(at: syntax.parent, traversedCatchClause: false)
+      } else {
+        return syntax
+      }
+    case .catchClause:
+      return lookupCatchNodeHelper(at: syntax.parent, traversedCatchClause: true)
+    case .tryExpr(let tryExpr):
+      if tryExpr.questionOrExclamationMark != nil {
+        return syntax
+      } else {
+        return lookupCatchNodeHelper(at: syntax.parent, traversedCatchClause: traversedCatchClause)
+      }
+    case .functionDecl, .accessorDecl, .initializerDecl, .deinitializerDecl, .closureExpr:
+      return syntax
+    case .exprList(let exprList):
+      if let tryExpr = exprList.first?.as(TryExprSyntax.self), tryExpr.questionOrExclamationMark != nil {
+        return Syntax(tryExpr)
+      }
+      return lookupCatchNodeHelper(at: syntax.parent, traversedCatchClause: traversedCatchClause)
+    default:
+      return lookupCatchNodeHelper(at: syntax.parent, traversedCatchClause: traversedCatchClause)
+    }
+  }
+
+  // MARK: - walkParentTree helper methods
+
+  /// Callect the first syntax node matching the collection type up to a function boundary.
+  private func walkParentTreeUpToFunctionBoundary<T: SyntaxProtocol>(
+    at syntax: Syntax?,
+    collect: T.Type
+  ) -> T? {
+    walkParentTreeUpToFunctionBoundary(at: syntax, collect: collect, stopWithFirstMatch: true).first
+  }
+
+  /// Callect syntax nodes matching the collection type up to a function boundary.
+  private func walkParentTreeUpToFunctionBoundary<T: SyntaxProtocol>(
+    at syntax: Syntax?,
+    collect: T.Type,
+    stopWithFirstMatch: Bool = false
+  ) -> [T] {
+    walkParentTree(
+      upTo: [
+        MemberBlockSyntax.self,
+        FunctionDeclSyntax.self,
+        InitializerDeclSyntax.self,
+        DeinitializerDeclSyntax.self,
+        AccessorDeclSyntax.self,
+        ClosureExprSyntax.self,
+      ],
+      at: syntax,
+      collect: collect,
+      stopWithFirstMatch: stopWithFirstMatch
+    )
+  }
+
+  /// Callect syntax nodes matching the collection type up until encountering one of the specified syntax nodes.
+  private func walkParentTree<T: SyntaxProtocol>(
+    upTo stopAt: [SyntaxProtocol.Type],
+    at syntax: Syntax?,
+    collect: T.Type,
+    stopWithFirstMatch: Bool = false
+  ) -> [T] {
+    guard let syntax, !stopAt.contains(where: { syntax.is($0) }) else { return [] }
+    if let matchedSyntax = syntax.as(T.self) {
+      if stopWithFirstMatch {
+        return [matchedSyntax]
+      } else {
+        return [matchedSyntax]
+          + walkParentTree(
+            upTo: stopAt,
+            at: syntax.parent,
+            collect: collect,
+            stopWithFirstMatch: stopWithFirstMatch
+          )
```

Comment:
FWIW, building up an array via `+` here is going to exhibit quadratic behavior, because we need to create a temporary array at each step. I think this recursive function can be replaced with an iterative one that will be more efficient, e.g.,

```swift
var matches: [T] = []
var nextSyntax: Syntax? = syntax // when 'syntax' goes away, replace with Syntax(self)
while let currentSyntax = nextSyntax {
  if stopAt.contains(where: { currentSyntax($0) }) {
    break
  }

  if let matchedSyntax = currentSyntax.as(T.Self) {
    matches.append(matchedSyntax)
    if stopWithFirstMatch {
      break
    }
  }

 nextSyntax = currentSyntax.parent
}
 
return matches
```

Reply:
Initially, I considered [implementing the method with an `inout` collector](https://github.com/swiftlang/swift-syntax/blob/9f2fab0a6d401f4ba2f7226bc2440b0d80914a15/Sources/SwiftLexicalScopes/Scope.swift#L57-L65) to benefit from tail recursion, but we scrapped that for simplicity. I think it’s a good idea to remove the recursion altogether for simplicity though. I replaced the implementation with your suggestion. Thanks!